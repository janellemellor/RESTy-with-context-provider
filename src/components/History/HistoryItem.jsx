import React from 'react';
import PropTypes from 'prop-types';

const HistoryItem = ({ url, method }) => (
  <section>
    <p>{method}</p>
    <p>{url}</p>
  </section>
); 

HistoryItem.propTypes = {
  url: PropTypes.string.isRequired,
  method: PropTypes.string.isRequired
};

export default HistoryItem;
