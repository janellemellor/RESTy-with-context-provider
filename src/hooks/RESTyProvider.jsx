import React, { createContext, useState, useContext } from 'react';
import { fetchRequest } from '../services/fetchRequest.js';

const RESTyContext = createContext();

// eslint-disable-next-line react/prop-types
export const RESTyProvider = ({ children }) => {
  const [url, setUrl] = useState('');
  const [method, setMethod] = useState('GET');
  const [jsonBody, setJsonBody] = useState('');
  const [response, setResponse] = useState({});

  const onChange = ({ target }) => {
    const setInputsFactory = {
      url: setUrl,
      method: setMethod,
      jsonBody: setJsonBody
    };

    setInputsFactory[target.name](target.value);
  };

  const onSubmit = (event) => {
    event.preventDefault();
    fetchRequest(url, method, jsonBody)
      .then(res => {
        setResponse(res);
      }); 
  };

  const context = {
    url, 
    method, 
    jsonBody, 
    response, 
    onChange, 
    onSubmit
  };
  
  return (
    <RESTyContext.Provider value={context} >
      {children}
    </RESTyContext.Provider>
  );
};

export const useRESTy = () => {
  const context = useContext(RESTyContext);
  return context;
};




