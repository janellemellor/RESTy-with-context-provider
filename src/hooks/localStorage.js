import { useState, useEffect } from 'react';

const fetchState = (key, initialValue) => localStorage.getItem(key) ? JSON.parse(localStorage.getItem(key)) : initialValue;

export const useLocalStorage = (key, initialValue) => {
  const [state, setState] = useState(fetchState(key, initialValue));

  useEffect(() => {
    localStorage.setItem(key, JSON.stringify(state));
  }, [state]);

  return [state, setState];
};
