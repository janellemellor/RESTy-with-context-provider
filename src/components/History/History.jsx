import React from 'react';
import HistoryItem from './HistoryItem';
import { useRESTy } from '../../hooks/RESTyProvider';

const History = () => {
  const { history } = useRESTy();
  
  const requestHistory = history.map(
    (item, i) => (
      <li key={i}>
        <HistoryItem {...item} />   
      </li>
    ));
  
  return (
    <ul>
      {requestHistory}
    </ul>    
  );
};

export default History;
