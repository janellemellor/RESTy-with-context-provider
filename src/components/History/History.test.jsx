import React from 'react';
import { shallow } from 'enzyme';
import History from './History';
import { useRESTy, RESTyProvider } from '../../hooks/RESTyProvider';

describe('History Component', () => {
  it('matches a snapshot', () => {
    const requestHistory = [
      {
        url: 'https://github.com/janellemellor',
        method: 'GET'
      },
      {
        url: 'https://github.com/janellemellor',
        method: 'POST'
      }  
    ];
    const wrapper = shallow(
      <RESTyProvider>
        <History history={requestHistory} />
      </RESTyProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });  
});

