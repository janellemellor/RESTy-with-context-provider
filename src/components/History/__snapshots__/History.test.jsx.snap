// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History Component matches a snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "history": Array [],
      "jsonBody": "",
      "method": "GET",
      "onChange": [Function],
      "onSubmit": [Function],
      "response": Object {},
      "url": "",
    }
  }
>
  <History
    history={
      Array [
        Object {
          "method": "GET",
          "url": "https://github.com/janellemellor",
        },
        Object {
          "method": "POST",
          "url": "https://github.com/janellemellor",
        },
      ]
    }
  />
</ContextProvider>
`;
